src/coreclr/vm/comdelegate.cpp

```diff
@@ -1941,7 +1908,57 @@ PCODE COMDelegate::GetInvokeMethodStub(EEImplMethodDesc* pMD)
         if (*pMD->GetSig() != (IMAGE_CEE_CS_CALLCONV_HASTHIS | IMAGE_CEE_CS_CALLCONV_DEFAULT))
             COMPlusThrow(kInvalidProgramException);
 
-        ret = COMDelegate::TheDelegateInvokeStub();
+        Stub *pStub = pClass->m_pMultiCastInvokeStub;
+        if (pStub == NULL)
+        {
+            MetaSig sig(pMD);
+
+            BOOL fReturnVal = !sig.IsReturnTypeVoid();
+
+            SigTypeContext emptyContext;
+            ILStubLinker sl(pMD->GetModule(), pMD->GetSignature(), &emptyContext, pMD, (ILStubLinkerFlags)(ILSTUB_LINKER_FLAG_STUB_HAS_THIS | ILSTUB_LINKER_FLAG_TARGET_HAS_THIS));
+
+            ILCodeStream *pCode = sl.NewCodeStream(ILStubLinker::kDispatch);
+            DWORD numMethodPtr = pCode->NewLocal(ELEMENT_TYPE_I);
+            
+            // Load the method pointer
+            pCode->EmitLoadThis();
+            pCode->EmitLDFLD(pCode->GetToken(CoreLibBinder::GetField(FIELD__DELEGATE__METHOD_PTR)));
+            pCode->EmitSTLOC(numMethodPtr);
+
+            // Load the target object
+            pCode->EmitLoadThis();
+            pCode->EmitLDFLD(pCode->GetToken(CoreLibBinder::GetField(FIELD__DELEGATE__TARGET)));
```

Comment:
I'm concerned around the GC safety around this:

After `_methodPtr` and `_target` are loaded, the lifetime of the delegate object should be end. Before PC pointing to the method, GC could consider the delegate and the method unreferenced, potentially collecting stubs or even the whole assembly.

Can this happen? Can it be resolved by making the local variable ELEMENT_TYPE_FNPTR?

Reply:
Yeah, good point. We may need GC.KeepAlive here for correctness.

`ELEMENT_TYPE_FNPTR` would not help. It is not GC tracked.

Reply:
This is not an easy problem indeed.

After the two fields are fetched, there's no register pointing to the delegate object. Calling `GC.KeepAlive` on the delegate will significantly degrade code quality.

`GC.KeepAlive` can't be called on a function pointer because it's not boxable. Assigning it to some another thing like `byref` would also confuse GC.